Make `seed!` and `extract_jacobian!` gpu-friendly

charleskawczynski opened the pull request:
Per @vchuravy's suggestion to open a new PR: Now that [JuliaLang/julia#14955](https://github.com/JuliaLang/julia/issues/14955) is resolved, this PR revives the work of #353 and #406.

charleskawczynski commented:
Interestingly, when trying
```julia
    ip = Iterators.product(ydual, 1:n)
    out_reshaped .= map(rc->partials(T, rc[1], rc[2]), ip)
```
I can see in the GPU stack trace:
```julia
ERROR: GPU compilation of kernel broadcast_kernel(CUDA.CuKernelContext, CuDeviceArray{Float32,2,1}, Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},typeof(identity),Tuple{Base.Broadcast.Extruded{Array{Float32,2},Tuple{Bool,Bool},Tuple{Int64,Int64}}}}, Int64) failed
KernelError: passing and using non-bitstype argument

Argument 4 to your kernel function is of type Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},typeof(identity),Tuple{Base.Broadcast.Extruded{Array{Float32,2},Tuple{Bool,Bool},Tuple{Int64,Int64}}}}, which is not isbits:
  .args is of type Tuple{Base.Broadcast.Extruded{Array{Float32,2},Tuple{Bool,Bool},Tuple{Int64,Int64}}} which is not isbits.
    .1 is of type Base.Broadcast.Extruded{Array{Float32,2},Tuple{Bool,Bool},Tuple{Int64,Int64}} which is not isbits.
      .x is of type Array{Float32,2} which is not isbits.
```
that the array in `Extruded` is not `similar` to the device array. I'm not familiar enough with `Extruded`, but maybe we can change initialization to use `similar` so that this solution is `isbits`?

Also, not sure who to ping here 😅

KristofferC commented on src/apiutils.jl:
```diff
@@ -55,36 +55,30 @@ end
-        j = i + offset
-        duals[j] = Dual{T,V,N}(x[j], seed)
-    end
+    i = 1:N
```
Perhaps it would be good to pick one way of writing this and use it consistently (regarding macros and whether to set `1:N` as a separate variable.

charleskawczynski replied:
Updated.

charleskawczynski commented:
Alright, I worked out the issue with `extract_jacobian!`, thanks @vchuravy! With this PR:
```
using ForwardDiff, BenchmarkTools
x = rand(1000);
cfg = ForwardDiff.GradientConfig(nothing, x);
@btime ForwardDiff.seed!($(cfg.duals), $x, $(cfg.seeds));
```
Results in 0 allocations, and
```
using CUDA, ForwardDiff, Test;
CUDA.allowscalar(false);

function f!(F, x)
  t = Tuple(x)
  F_nt = ntuple(Val(length(F))) do i
      i==1 && (F_i = (t[1] + 3) * (t[2]^3 - 7) + 18)
      i==2 && (F_i = sin(t[2] * exp(t[1]) - 1))
      F_i
  end
  F .= F_nt
end;

x_initial = CUDA.CuArray(Float32[0.1; 1.2]);
# x_initial = Float32[0.1; 1.2];
J = similar(x_initial, (length(x_initial),length(x_initial)));
F = similar(x_initial);
Ja =[-5.272    13.392
  1.25627   1.04689]
ForwardDiff.jacobian!(J, f!, F, x_initial);
@test all(Array(J) .≈ Ja)
```
passes.

KristofferC commented on src/apiutils.jl:
```diff
@@ -55,36 +55,30 @@ end
-        duals[j] = Dual{T,V,N}(x[j], seed)
-    end
+    j = (1:N) .+ offset
+    duals[j] .= Dual{T,V,N}(x[j], seed)
```
Shouldn't this be a view onto `x`?

charleskawczynski replied:
Ah, yes, and I was missing broadcasts and `Ref`s. Fixed!

charleskawczynski commented:
Hmmm, I can't reproduce the CI errors locally 🤔. CI passes for me locally both on my Windows and Mac machines.

simonbyrne commented:
This would be a good candidate for the julialang GPU CI.

charleskawczynski commented:
bump

oxinabox commented on src/apiutils.jl:
```diff
@@ -55,36 +55,30 @@ end
-    end
+    i = 1:chunksize
+    j = i .+ offset
+    duals[j] .= Dual{T,V,N}.(view(x, j), seeds[i])
```
I would be tempted to write `seed_inds` and `dual_inds` rather than `i` and `j` because the single letter makes me think it is a scalar.
Similar above where there is 1 variable i would be tempted to call it `inds` rather than `i` or `j` for same reason.

or `irange` like used below

oxinabox commented on src/jacobian.jl:
```diff
@@ -111,9 +111,10 @@ end
-    for col in 1:size(out_reshaped, 2), row in 1:size(out_reshaped, 1)
-        out_reshaped[row, col] = partials(T, ydual[row], col)
-    end
+    ydual_reshaped = reshape(ydual, length(ydual))
```
```suggestion
    ydual_reshaped = vec(ydual)
```

oxinabox commented on src/jacobian.jl:
```diff
@@ -123,13 +124,13 @@ function extract_jacobian!(::Type{T}, result::MutableDiffResult, ydual::Abstract
 end
 
 function extract_jacobian_chunk!(::Type{T}, result, ydual, index, chunksize) where {T}
+    ydual_reshaped = reshape(ydual, length(ydual))
```
```suggestion
    ydual_reshaped = vec(ydual)
```

oxinabox reviewed (approved):
LGTM, though some small comments about naming thing and using `vec` for reshaping to a vector.

This *might* be a tiny bit slower than it is now for `extract_jacobian_chunk!`,
but i think it probably optimized out, at least in 1.5, and that is not perforance critical anyway, so if this gets us GPU support than that is a win.

Please bump the version number as a non-breaking change.


Strong agreement re: setting up JuliaGPU tests, but I won't block the PR over that.

Once these things are addressed I will merge.

charleskawczynski commented:
> LGTM, though some small comments about naming thing and using vec for reshaping to a vector.

I changed to `vec`, looks like it still runs fine on the GPU.

> Please bump the version number as a non-breaking change.

I bumped `version = "0.10.12"` ➡️  `version = "0.10.13"`, this is the version bump you mean, right?

KristofferC commented:
How did the benchmark show 0 allocations when views and broadcasting were missed in some places? Where those not benchmarked? Would be good to ensure all the stuff changed here has been benchmarked and confirmed not to allocate.

charleskawczynski commented:
> How did the benchmark show 0 allocations when views and broadcasting were missed in some places? Where those not benchmarked? Would be good to ensure all the stuff changed here has been benchmarked and confirmed not to allocate.

Where were views/broadcasts missing? I agree that adding a test would be nice. I'll see if I can add that somewhere.

KristofferC commented:
I though e.g https://github.com/JuliaDiff/ForwardDiff.jl/pull/472#discussion_r520373059 would have caused allocations.

charleskawczynski commented:
> I though e.g [#472 (comment)](https://github.com/JuliaDiff/ForwardDiff.jl/pull/472#discussion_r520373059) would have caused allocations.

I must have had the view in the `seed!(duals::AbstractArray{Dual{T,V,N}}, x, seeds::NTuple{N,Partials{N,V}}) where {T,V,N}` signature but not in `seed!(duals::AbstractArray{Dual{T,V,N}}, x, index, seed::Partials{N,V} = zero(Partials{N,V})) where {T,V,N}`, where you commented-- I'm not sure since I squashed that commit.

For good measure, I added `BenchmarkTools` to the test environment and added tests to ensure that we're getting zero allocations for all three `seed!` definitions.

charleskawczynski commented:
Ah, glad I added the tests, I just realized that I missed one method (the `seed!` with `NTuple` seeds and `index`), and that one happens to be failing. Will update soon.

charleskawczynski commented:
Alright, all of the `seed!` methods are non-allocating and are now tested to be non-allocating.

charleskawczynski commented:
Ok, I also squashed. I think this should be ready to go.

oxinabox commented:
Allocation tests are failing on julia 1.4 and 1.0.
You probably want to only run that file on 1.5 and above.

KristofferC commented on test/AllocationsTest.jl:
```diff
@@ -0,0 +1,27 @@
+    x = rand(1000)
+    cfg = ForwardDiff.GradientConfig(nothing, x)
+
+    balloc = @ballocated ForwardDiff.seed!($(cfg.duals), $x, $(cfg.seeds))
```
Doesn't just `@allocated` work?

charleskawczynski replied:
I tried, but I wasn't able to interpolate with `@allocated`, so it results in non-zero allocations.

charleskawczynski replied:
Actually, it looks like we can use `@allocated`, we'll just need to duplicate the calls and remove the `getproperty`/`getindex` in the calls:
```julia
@testset "Test seed! allocations" begin
    x = rand(1000)
    cfg = ForwardDiff.GradientConfig(nothing, x)
    duals = cfg.duals
    seeds = cfg.seeds
    seed = cfg.seeds[1]

    balloc = @allocated ForwardDiff.seed!(duals, x, seeds)
    balloc = @allocated ForwardDiff.seed!(duals, x, seeds)
    @test balloc == 0

    balloc = @allocated ForwardDiff.seed!(duals, x, seed)
    balloc = @allocated ForwardDiff.seed!(duals, x, seed)
    @test balloc == 0

    index = 1
    balloc = @allocated ForwardDiff.seed!(duals, x, index, seeds)
    balloc = @allocated ForwardDiff.seed!(duals, x, index, seeds)
    @test balloc == 0

    index = 1
    balloc = @allocated ForwardDiff.seed!(duals, x, index, seed)
    balloc = @allocated ForwardDiff.seed!(duals, x, index, seed)
    @test balloc == 0
end
```
Let me know which one is preferred

oxinabox replied:
I think that originally (in e.g. 1.0) julia tried to excluded compilation from the output of `@allocated`, but that proved unreliable, and so that was removed in 1.4 (ish?).
In response to that I added `@ballocated` to `BenchmarkTools` that just pulled the allocations out of the fastest run according to `@benchmark` (which will naturally not be the one that included the timing).

I think `@ballocated` is nicer than duplicating personally, but I will defer to others.

charleskawczynski replied:
I agree that `@ballocated` seems nicer, but I'll also defer to others as this is my first PR to this repo.

KristofferC replied:
`@ballocated` doesn't really make sense to me. You run the function thousands of time and then pick out the lowest one. For functions that don't allocate, that will be the second run. So why run the function 998 more times?

charleskawczynski replied:
Ok, I'll switch to the code snippet above

charleskawczynski replied:
Updated.

charleskawczynski commented:
> Allocation tests are failing on julia 1.4 and 1.0.
> You probably want to only run that file on 1.5 and above.
> 
> The failure on nightly is unrelated and is fixed by
> #476

Ah, good point. I'll update to test on 1.5 and above.

charleskawczynski commented:
Bump

charleskawczynski commented:
Rebased and re-bumped the version.
